### PR TITLE
feat: support overwrite X-Forwarded-* 

### PR DESCRIPTION
**Pull request by rapiz1**
### What this PR does / why we need it:

Support `X-Forwarded-For`,`X-Forwarded-Host`,`X-Forwarded-Port` rewriting for `proxy-rewrite` plugin.

Resolves https://github.com/apache/apisix/issues/5080

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/init.lua [hunk 1]
```diff
@@ -269,9 +269,18 @@ end
 local function set_upstream_headers(api_ctx, picked_server)
     set_upstream_host(api_ctx, picked_server)
 
-    local hdr = core.request.header(api_ctx, "X-Forwarded-Proto")
-    if hdr then
-        api_ctx.var.var_x_forwarded_proto = hdr
+    local proxy_headers = {
```
Comment (tokers) [inline apisix/init.lua:272]: It can be defined as a module variable.

Comment (rapiz1) [reply apisix/init.lua]: done

File: apisix/init.lua [hunk 2]
```diff
@@ -266,12 +266,21 @@ local function set_upstream_host(api_ctx, picked_server)
...
+    var_x_forwarded_port = 'X-Forwarded-Port',
+}
+
 local function set_upstream_headers(api_ctx, picked_server)
     set_upstream_host(api_ctx, picked_server)
 
-    local hdr = core.request.header(api_ctx, "X-Forwarded-Proto")
-    if hdr then
-        api_ctx.var.var_x_forwarded_proto = hdr
+    for ngx_key, hdr_key in pairs(upstream_proxy_headers) do
```
Comment (spacewander) [inline apisix/init.lua:279]: We can also remove https://github.com/apache/apisix/blob/a0efeca194bcca58e62d73c0a19464f480a1b7c9/apisix/cli/ngx_tpl.lua#L600 and https://github.com/apache/apisix/blob/a0efeca194bcca58e62d73c0a19464f480a1b7c9/t/APISIX.pm#L697

File: apisix/init.lua [hunk 3]
```diff
@@ -266,12 +266,21 @@ local function set_upstream_host(api_ctx, picked_server)
 end
 
 
+local upstream_proxy_headers = {
+    var_x_forwarded_proto = 'X-Forwarded-Proto',
+    var_x_forwarded_for = 'X-Forwarded-For',
```
Comment (spacewander) [inline apisix/init.lua:271]: For X-Forwarded-For, we need to append the client's real-ip to it, see https://github.com/apache/apisix/blob/a0efeca194bcca58e62d73c0a19464f480a1b7c9/t/APISIX.pm#L698

Comment (rapiz1) [reply apisix/init.lua:271]: I'm uncertain about this. `proxy-rewrite` plugin simply overwrites the original value for other headers. If `X-Forward-For` is specialized, the behavior is not consistent, which may be a confusion for the users.

File: apisix/init.lua (hunk 2 above)
Comment (rapiz1) [reply apisix/init.lua:279]: Can you please explain why the requested changes are needed?

Comment (spacewander) [reply apisix/init.lua:279]: The same feature is implemented in your PR, except for the  X-Forwarded-For one. After your change, there won't be client IP appending to the http_x_forwarded_for  when the header is not rewritten. We should add test for it and fix it.

**Comment by jiayx** [comment]
Are there any updates?

**Comment by juzhiyuan** [comment]
Hi @jiayx, AFAIK, @rapiz1 is busy with other things 🤔 Would like to finish this PR?

**Comment by jiayx** [comment]
Okay, I'll try my best.

**Pull request closed without merging**